In [1]:
import pandas as pd

df = pd.read_csv('/content/country_vaccinations.csv')
display(df.head())

country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  
0  World Health Organization  https://covid19.who.int/  
1  World Health Organization  https://covid19.who.int/  
2  World Health Organization  https://covid19.who.int/  
3  World Health Organization  https://covid19.who.int/  
4  World Health Organization  https://covid19.who.int/

## Preparar y Visualizar Estructura del DataFrame

### Subtask:
Mostrar la estructura y los tipos de datos de cada columna del DataFrame usando `df.info()`. Luego, convertir la columna de fechas a un tipo de datos `datetime64` para permitir operaciones de tiempo adecuadas.


In [3]:
df.info()
df['date'] = pd.to_datetime(df['date'])
print("Date column converted to datetime type.")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86512 entries, 0 to 86511
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              86512 non-null  object        
 1   iso_code                             86512 non-null  object        
 2   date                                 86512 non-null  datetime64[ns]
 3   total_vaccinations                   43607 non-null  float64       
 4   people_vaccinated                    41294 non-null  float64       
 5   people_fully_vaccinated              38802 non-null  float64       
 6   daily_vaccinations_raw               35362 non-null  float64       
 7   daily_vaccinations                   86213 non-null  float64       
 8   total_vaccinations_per_hundred       43607 non-null  float64       
 9   people_vaccinated_per_hundred        41294 non-null  float64       
 10  people_ful

## Vacunas Aplicadas por Compañía

### Subtask:
Calcular la cantidad total de vacunas aplicadas por cada compañía. Esto se logrará agrupando el DataFrame por la columna `vaccines` y sumando las `total_vaccinations`.


In [5]:
total_vaccines_by_company = df.groupby('vaccines')['total_vaccinations'].sum().sort_values(ascending=False)
print("Total vaccines applied by company:")
display(total_vaccines_by_company)

Total vaccines applied by company:


vaccines
CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, Sinovac, ZF2001     7.094527e+11
Covaxin, Oxford/AstraZeneca, Sputnik V                           3.234403e+11
Johnson&Johnson, Moderna, Pfizer/BioNTech                        1.617576e+11
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                     1.153913e+11
Johnson&Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech    9.962579e+10
                                                                     ...     
Pfizer/BioNTech, Sinopharm/Beijing, Sputnik V                    6.915494e+06
Oxford/AstraZeneca, Pfizer/BioNTech, Sputnik V                   6.840405e+06
Moderna                                                          6.531798e+06
Johnson&Johnson, Moderna                                         4.984467e+06
Pfizer/BioNTech, Sputnik V                                       3.863380e+06
Name: total_vaccinations, Length: 84, dtype: float64

## Total de Vacunas Aplicadas Mundialmente

### Subtask:
Determinar la cantidad total de vacunas aplicadas en todo el mundo sumando la columna `total_vaccinations`.


In [7]:
total_worldwide_vaccinations = df['total_vaccinations'].sum()
print(f"Total worldwide vaccinations: {total_worldwide_vaccinations:.2f}")

Total worldwide vaccinations: 2002854013761.00


## Promedio de Vacunas por País

### Subtask:
Calcular el promedio de vacunas aplicadas por cada país, agrupando por `country` y obteniendo el promedio de `total_vaccinations`.


In [8]:
avg_vaccines_per_country = df.groupby('country')['total_vaccinations'].mean().sort_values(ascending=False)
print("Average total vaccinations per country:")
display(avg_vaccines_per_country)

Average total vaccinations per country:


country
China               1.842734e+09
India               7.557016e+08
United States       3.291165e+08
Brazil              1.838771e+08
Indonesia           1.192998e+08
                        ...     
Montserrat          2.740229e+03
Falkland Islands    2.360500e+03
Niue                1.842500e+03
Tokelau             9.642500e+02
Pitcairn            6.960000e+01
Name: total_vaccinations, Length: 223, dtype: float64

## Vacunas Aplicadas en una Fecha Específica

### Subtask:
Obtener la cantidad total de vacunas aplicadas en todo el mundo el día 29 de enero de 2021, filtrando el DataFrame por la fecha y sumando las `daily_vaccinations`.


In [9]:
vaccinations_jan_29 = df[df['date'] == '2021-01-29']['daily_vaccinations'].sum()
print(f"Total worldwide daily vaccinations on 2021-01-29: {vaccinations_jan_29:.2f}")

Total worldwide daily vaccinations on 2021-01-29: 4884052.00


## Crear DataFrame 'conDiferencias'

### Subtask:
Crear un nuevo DataFrame llamado `conDiferencias` que contenga todos los datos originales, más una nueva columna llamada `diferencias`. Esta columna contendrá la diferencia entre las columnas `daily_vaccinations` y `daily_vaccinations_raw`.


In [10]:
conDiferencias = df.copy()
conDiferencias['diferencias'] = conDiferencias['daily_vaccinations'] - conDiferencias['daily_vaccinations_raw']
print("DataFrame 'conDiferencias' created with new 'diferencias' column:")
display(conDiferencias.head())

DataFrame 'conDiferencias' created with new 'diferencias' column:


country iso_code       date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG 2021-02-22                 0.0                0.0   
1  Afghanistan      AFG 2021-02-23                 NaN                NaN   
2  Afghanistan      AFG 2021-02-24                 NaN                NaN   
3  Afghanistan      AFG 2021-02-25                 NaN                NaN   
4  Afghanistan      AFG 2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  diferencias  
0  World Health Organization  https://covid19.who.int/          NaN  
1  World Health Organization  https://covid19.who.int/          NaN  
2  World Health Organization  https://covid19.who.int/          NaN  
3  World Health Organization  https://covid19.who.int/          NaN  
4  World Health Organization  https://covid19.who.int/          NaN

## Período de Tiempo de los Registros

### Subtask:
Calcular el período de tiempo (días, meses, años) entre el registro con la fecha más reciente y el registro con la fecha más antigua en el DataFrame.


In [11]:
min_date = df['date'].min()
max_date = df['date'].max()
time_span = max_date - min_date

print(f"Earliest record date: {min_date}")
print(f"Latest record date: {max_date}")
print(f"Time span of records: {time_span}")

Earliest record date: 2020-12-02 00:00:00
Latest record date: 2022-03-29 00:00:00
Time span of records: 482 days 00:00:00


## Crear DataFrame 'conCantidad'

### Subtask:
Crear un nuevo DataFrame llamado `conCantidad` que incluya todos los datos originales y una nueva columna llamada `canVac`. Esta columna contendrá la cantidad de tipos de vacunas utilizadas en cada registro, obtenida dividiendo la columna `vaccines` por la coma (`,`) y contando los elementos.


In [12]:
conCantidad = df.copy()
conCantidad['canVac'] = conCantidad['vaccines'].apply(lambda x: len(str(x).split(', ')))
print("DataFrame 'conCantidad' created with new 'canVac' column:")
display(conCantidad.head())

DataFrame 'conCantidad' created with new 'canVac' column:


country iso_code       date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG 2021-02-22                 0.0                0.0   
1  Afghanistan      AFG 2021-02-23                 NaN                NaN   
2  Afghanistan      AFG 2021-02-24                 NaN                NaN   
3  Afghanistan      AFG 2021-02-25                 NaN                NaN   
4  Afghanistan      AFG 2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  canVac  
0  World Health Organization  https://covid19.who.int/       4  
1  World Health Organization  https://covid19.who.int/       4  
2  World Health Organization  https://covid19.who.int/       4  
3  World Health Organization  https://covid19.who.int/       4  
4  World Health Organization  https://covid19.who.int/       4

## Crear DataFrame 'antes20'

### Subtask:
Crear un nuevo DataFrame llamado `antes20` que contenga todos los registros del DataFrame original `df` cuya fecha sea anterior al 20 de diciembre de 2020.



In [13]:
antes20 = df[df['date'] < '2020-12-20']
print("DataFrame 'antes20' created with records before 2020-12-20:")
display(antes20.head())

DataFrame 'antes20' created with records before 2020-12-20:


country iso_code       date  total_vaccinations  people_vaccinated  \
13403  Canada      CAN 2020-12-14                 5.0                5.0   
13404  Canada      CAN 2020-12-15               727.0              727.0   
13405  Canada      CAN 2020-12-16              3025.0             3025.0   
13406  Canada      CAN 2020-12-17              7279.0             7279.0   
13407  Canada      CAN 2020-12-18             11296.0            11296.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
13403                      NaN                     NaN                 NaN   
13404                      NaN                   722.0               722.0   
13405                      NaN                  2298.0              1510.0   
13406                      NaN                  4254.0              2425.0   
13407                      NaN                  4017.0              2823.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
13403                            0.00                           0.00   
13404                            0.00                           0.00   
13405                            0.01                           0.01   
13406                            0.02                           0.02   
13407                            0.03                           0.03   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
13403                                  NaN                             NaN   
13404                                  NaN                            19.0   
13405                                  NaN                            40.0   
13406                                  NaN                            64.0   
13407                                  NaN                            74.0   

                                                vaccines  \
13403  Johnson&Johnson, Moderna, Oxford/AstraZeneca, ...   
13404  Johnson&Johnson, Moderna, Oxford/AstraZeneca, ...   
13405  Johnson&Johnson, Moderna, Oxford/AstraZeneca, ...   
13406  Johnson&Johnson, Moderna, Oxford/AstraZeneca, ...   
13407  Johnson&Johnson, Moderna, Oxford/AstraZeneca, ...   

                                             source_name  \
13403  Official data from provinces via covid19tracke...   
13404  Official data from provinces via covid19tracke...   
13405  Official data from provinces via covid19tracke...   
13406  Official data from provinces via covid19tracke...   
13407  Official data from provinces via covid19tracke...   

                                          source_website  
13403  https://covid19tracker.ca/vaccinationtracker.html  
13404  https://covid19tracker.ca/vaccinationtracker.html  
13405  https://covid19tracker.ca/vaccinationtracker.html  
13406  https://covid19tracker.ca/vaccinationtracker.html  
13407  https://covid19tracker.ca/vaccinationtracker.html

## Crear DataFrame 'pfizer'

### Subtask:
Crear un nuevo DataFrame llamado `pfizer` que contenga todos los registros del DataFrame original `df` donde la columna `vaccines` indique que se utilizó la vacuna 'Pfizer'.



In [14]:
pfizer = df[df['vaccines'].str.contains('Pfizer', na=False)]
print("DataFrame 'pfizer' created with records where 'Pfizer' vaccine was used:")
display(pfizer.head())

DataFrame 'pfizer' created with records where 'Pfizer' vaccine was used:


country iso_code       date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG 2021-02-22                 0.0                0.0   
1  Afghanistan      AFG 2021-02-23                 NaN                NaN   
2  Afghanistan      AFG 2021-02-24                 NaN                NaN   
3  Afghanistan      AFG 2021-02-25                 NaN                NaN   
4  Afghanistan      AFG 2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  
0  World Health Organization  https://covid19.who.int/  
1  World Health Organization  https://covid19.who.int/  
2  World Health Organization  https://covid19.who.int/  
3  World Health Organization  https://covid19.who.int/  
4  World Health Organization  https://covid19.who.int/

## Exportar DataFrames a Excel

### Subtask:
Exportar los DataFrames `conDiferencias`, `conCantidad`, `antes20`, y `pfizer` a un archivo Excel llamado `resultadosReto.xlsx`. Cada DataFrame debe estar en una hoja separada dentro del archivo Excel.



In [15]:
output_file = 'resultadosReto.xlsx'
with pd.ExcelWriter(output_file) as writer:
    conDiferencias.to_excel(writer, sheet_name='conDiferencias', index=False)
    conCantidad.to_excel(writer, sheet_name='conCantidad', index=False)
    antes20.to_excel(writer, sheet_name='antes20', index=False)
    pfizer.to_excel(writer, sheet_name='pfizer', index=False)
print(f"DataFrames exported to {output_file}")

DataFrames exported to resultadosReto.xlsx


## Resumen de los Hallazgos

### Sumario:
En esta fase del análisis, se ha explorado un conjunto de datos de vacunaciones por país, realizando diversas transformaciones y cálculos para extraer información clave. Los principales hallazgos son los siguientes:

1.  **Estructura y Preparación de Datos:**
    *   Se inspeccionó la estructura del DataFrame `df`, confirmando los tipos de datos de cada columna. La columna 'date' fue convertida exitosamente al tipo `datetime64` para facilitar operaciones temporales.

2.  **Vacunas por Compañía:**
    *   Se calculó la cantidad total de vacunas aplicadas por cada combinación de compañías de vacunas, mostrando que ciertas combinaciones tienen una contribución significativamente mayor a la vacunación global.

3.  **Vacunación Mundial:**
    *   La cantidad total de vacunas aplicadas mundialmente es de aproximadamente **2,002,854,013,761.00**.

4.  **Promedio por País:**
    *   Se determinó el promedio de vacunas aplicadas por cada país, identificando a China, India y Estados Unidos como los países con el mayor promedio de vacunaciones totales.

5.  **Vacunación en Fecha Específica:**
    *   El 29 de enero de 2021, el total de vacunaciones diarias a nivel mundial fue de **4,884,052.00**.

6.  **DataFrame `conDiferencias`:**
    *   Se creó `conDiferencias`, que incluye una columna `diferencias` mostrando la discrepancia entre `daily_vaccinations` y `daily_vaccinations_raw`. Esto puede ser útil para identificar ajustes o métodos de cálculo en los datos diarios.

7.  **Período de Tiempo:**
    *   El rango de fechas cubierto por los registros va desde el **2 de diciembre de 2020** hasta el **29 de marzo de 2022**, abarcando un período total de **482 días**.

8.  **DataFrame `conCantidad`:**
    *   Se generó `conCantidad` con una nueva columna `canVac`, que indica el número de tipos de vacunas utilizadas en cada registro, proporcionando una visión de la diversidad de vacunas implementadas.

9.  **DataFrame `antes20`:**
    *   El DataFrame `antes20` se creó con registros de vacunación anteriores al 20 de diciembre de 2020, lo que permite un análisis específico del inicio de las campañas de vacunación.

10. **DataFrame `pfizer`:**
    *   El DataFrame `pfizer` fue filtrado para incluir solo los registros donde se utilizó la vacuna 'Pfizer', facilitando un análisis centrado en la distribución y uso de esta vacuna.

11. **Exportación de Datos:**
    *   Finalmente, los DataFrames `conDiferencias`, `conCantidad`, `antes20`, y `pfizer` fueron exportados exitosamente a un archivo Excel llamado `resultadosReto.xlsx`, con cada DataFrame en una hoja separada para su posterior revisión y análisis.

## Final Task

### Subtask:
Resumir los hallazgos principales y las acciones realizadas en este análisis de datos.
